mirac_network/mirac-broker.cpp

```diff
@@ -103,11 +111,30 @@ gboolean MiracBroker::connect_cb (gint fd, GIOCondition condition)
         g_unix_fd_add(connection_->GetHandle(), G_IO_IN, receive_cb, this);
         on_connected();
     } catch (std::exception &x) {
-        g_warning("exception: %s", x.what());
+        gdouble elapsed = 1000 * g_timer_elapsed(connect_timer_, NULL);
+        if (elapsed + connect_wait_ > connect_timeout_) {
+            on_connect_timeout();
+        } else {
+            connect_wait_id_ = g_timeout_add (connect_wait_, try_connect, this);
+        }
     }
     return G_SOURCE_REMOVE;
 }
 
+void MiracBroker::try_connect()
+{
+    g_message("Trying to connect...");
+
+    connect_wait_id_ = 0;
+    network_.reset(new MiracNetwork());
+
+    if (network_->Connect(peer_address_.c_str(), peer_port_.c_str())) {
```

Comment:
Network::Connect() still takes char* ... but you are right, we should change that if the owner is already using std::string anyway and the rest of the API is already using std::string: I'll add a commit